tfjs-backend-webgpu/src/matmul_test.ts

```diff
@@ -1005,4 +1023,773 @@ describeWebGPU('matmul', () => {
+        ],
+        [4, 5, 2, 3]);
+    const b = tf.tensor4d(
+        [
+          -4, -3, -2, -6, 6,  -1, -4, -1, 7,  -4, 8,  -9,  -9, 0,   -1,
+          -4, -6, -7, -3, -4, -7, 6,  -8, 1,  -2, 1,  -1,  -3, 8,   -5,
+          9,  -2, 5,  9,  -2, 2,  -5, -5, -8, -1, -2, -3,  -2, -10, 6,
+          -3, 0,  1,  6,  7,  1,  2,  -4, -5, 2,  -5, -7,  9,  3,   -6,
+          6,  4,  -4, 6,  10, -3, -2, 8,  10, -8, 10, -1,  -9, -7,  -8,
+          -3, 1,  1,  -2, -9, -7, -6, -1, 0,  7,  -9, -7,  -5, 0,   -4,
+          -4, -7, 2,  4,  6,  6,  -4, -6, -8, 3,  -8, -9,  6,  9,   -4,
+          1,  -1, 0,  8,  9,  0,  -5, 3,  -1, 5,  0,  -10, 7,  -2,  6
+        ],
+        [4, 5, 2, 3]);
+
+    const transposeA = true;
+    const transposeB = false;
+    const c = tf.matMul(a, b, transposeA, transposeB);
+
+    expectArraysClose(await c.data(), [
+      38,  -24, 9,   -30, 9,   -9,  -74,  39,  -19,  8,    11,  -30, 56,  -67,
+      46,  -40, 71,  -74, 82,  42,  55,   -50, 6,    1,    60,  -18, -13, -15,
+      -52, -61, 81,  -52, 59,  -15, 76,   43,  34,   -56,  38,  0,   26,  -14,
+      -15, 1,   -4,  153, -34, 61,  -135, 30,  -48,  135,  -30, 60,  38,  36,
+      58,  40,  45,  71,  1,   2,   3,    24,  90,   -56,  -10, 40,  -18, 6,
+      -30, 14,  34,  65,  27,  24,  -29,  -44, -46,  -3,   35,  -21, 27,  48,
+      20,  52,  32,  35,  -11, -46, -12,  22,  13,   30,   2,   -23, -54, -48,
+      34,  16,  -42, -39, -26, 82,  89,   76,  -84,  30,   9,   27,  30,  -21,
+      -43, -48, 60,  20,  24,  -78, -91,  -63, -12,  24,   21,  28,  48,  35,
+      -6,  27,  33,  53,  -81, -71, 61,   -27, 11,   -48,  -82, 8,   -12, -19,
+      -10, -48, -81, 0,   13,  32,  41,   0,   -100, -120, 16,  124, 152, 45,
+      60,  -28, 24,  21,  -12, -14, -16,  8,   9,    -33,  5,   -12, -48, 4,
+      8,   9,   0,   -31, 16,  -98, -9,   4,   -22,  38,   2,   -96
+    ]);
+  });
+
+  it('A^t x B^t', async () => {
+    const a = tf.tensor3d(
+        [
+          -5, -5, -6, 8, -2, -8, 4, -7, -6, -9, -1, 3,  7,  -2, 5,
+          -6, 3,  8,  7, -8, 1,  4, -4, 6,  4,  -4, -9, -5, 2,  -2
+        ],
+        [5, 3, 2]);
+    const b = tf.tensor3d(
+        [
+          -8, -4, -1, 0,  -7, 0, 3,  3,  6,   2,  -1, 8, -4, 9, -6,
+          5,  8,  9,  -9, 7,  0, -1, -1, -10, -7, 3,  4, 6,  3, -4
+        ],
+        [5, 2, 3]);
+
+    const transposeA = true;
+    const transposeB = true;
+    const c = tf.matMul(a, b, transposeA, transposeB);
+    expectArraysClose(await c.data(), [
+      66,  42, 16,  -56, -12, 6,   -30, 19,  -1, 102,
+      -94, 14, -56, 32,  100, -56, -47, -11, 5,  -31
+    ]);
+  });
+};
+
+for (let i = 0; i < MatMulProgramType.MatMULMax; i++) {
+  describeWithFlags(`matmul with program type ${i}`, MATMUL_ENVS, matmulTest);
+  // Skip MatMulSplitKProgram since it doesn't support batch > 1;
+  if (i !== MatMulProgramType.MatMulSplitKProgram) {
+    describeWithFlags(
+        `matmulBatch with program type ${i}`, MATMUL_ENVS, matmulBatchTest);
+  }
+}
```

Comment:
Another approach is to use a different environment for each MatMulProgramType. If you want to test all the WebGPU tests against the different MatMul programs, then this other approach is likely better. If you only want to test the matmul tests with different matmul programs, then your approach looks good, but skip to the end of this comment for some potential improvements.

To use a different environment for each `MatMulProgramType`, you can [register a test environment](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/jasmine_util.ts#L240-L255) for each value of the `MatMulProgramType` enum:
```ts
// backend_webgpu_test_registry.ts
for (let i = 0; i < MatMulProgramType.MatMULMax; i++) {
  registerTestEnv({
    name: `webgpu ${MatMulProgramType[i]}`,
    backendName: 'webgpu',
    flags: {
      'WEBGPU_TESTED_MATMUL': i,
    },
    isDataSync: false,
  }
}
```

Then, `describeWithFlags` will [run the specs on all of the registered environments](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/jasmine_util.ts#L218-L220) (as long as the environment satisfies the constraints).

Then, for matmul tests, you would just need to write `describeWithFlags('matmulBatch', ALL_ENVS, () => { /* tests */ })`.

However, this approach will also run all the other tests with these MatMul environments, meaning all the tests would run 5 times. If this is not what you want, you can create a constraint that requires `WEBGPU_TESTED_MATMUL` to be -1 and apply it to all the other tests:
```ts
export const DEFAULT_MATMUL_ENVS: Constraints = {
  flags: {'WEBGPU_TESTED_MATMUL': -1}
};
```
However, this may require a lot of refactoring. 

Custom environments are more suited to flags that affect all the tests instead of just a small subset of them. Another approach which is better for flags that have a small effect (as appears to be the case here) is to just set the flag in a `beforeAll` and then reset it in `afterAll` (i.e. what you've done here).

```ts
for (let i = 0; i < MatMulProgramType.MatMULMax; i++) {
  describeWithFlags(`matmul ${MatMulProgramType[i]}`, ALL_ENVS, () => {
    let savedMatmulFlag: number;
    beforeAll(() => {
      savedMatmulFlag = tf.env().get('WEBGPU_TESTED_MATMUL') as number;
      tf.env().set('WEBGPU_TESTED_MATMUL', i);
    });

    afterAll(() => {
      tf.env().set('WEBGPU_TESTED_MATMUL, savedMatmulFlag);
    });

    ...
  }
}
```
With this approach, you don't need to create a custom `Constraint` or custom `TestEnv`s. Everything is controlled by the `for` loop and `beforeAll` / `afterAll`. If you don't need to run all the WebGPU tests with different values of `MatMulProgramType` and only want to run the MatMul tests with them, then I think this is the best approach.